flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -428,33 +434,66 @@ public int getIndex() {
             this.jobID = jobID;
             this.blobWriter = blobWriter;
+            this.offloadShuffleDescriptorsThreshold = offloadShuffleDescriptorsThreshold;
         }
 
         public MaybeOffloaded<ShuffleDescriptorAndIndex[]> serializeAndTryOffloadShuffleDescriptor(
-                ShuffleDescriptorAndIndex[] shuffleDescriptors) throws IOException {
+                ShuffleDescriptorAndIndex[] shuffleDescriptors, int numConsumer)
+                throws IOException {
 
             final CompressedSerializedValue<ShuffleDescriptorAndIndex[]> compressedSerializedValue =
                     CompressedSerializedValue.fromObject(shuffleDescriptors);
 
             final Either<SerializedValue<ShuffleDescriptorAndIndex[]>, PermanentBlobKey>
                     serializedValueOrBlobKey =
-                            BlobWriter.tryOffload(compressedSerializedValue, jobID, blobWriter);
+                            shouldOffload(shuffleDescriptors, numConsumer)
+                                    ? BlobWriter.offloadWithException(
+                                            compressedSerializedValue, jobID, blobWriter)
+                                    : Either.Left(compressedSerializedValue);
 
             if (serializedValueOrBlobKey.isLeft()) {
                 return new TaskDeploymentDescriptor.NonOffloaded<>(serializedValueOrBlobKey.left());
             } else {
                 return new TaskDeploymentDescriptor.Offloaded<>(serializedValueOrBlobKey.right());
             }
         }
+
+        /**
+         * Determine whether shuffle descriptors should be offloaded by blob server.
+         *
+         * @param shuffleDescriptorsToSerialize shuffle descriptors to serialize
+         * @param numConsumers how many consumers this serialized shuffle descriptor should be sent
+         * @return whether shuffle descriptors should be offloaded by blob server
+         */
+        private boolean shouldOffload(
+                ShuffleDescriptorAndIndex[] shuffleDescriptorsToSerialize, int numConsumers) {
+            return shuffleDescriptorsToSerialize.length * numConsumers
+                    >= offloadShuffleDescriptorsThreshold;
+        }
     }
+
+    @Experimental
+    @Documentation.ExcludeFromDocumentation(
+            "this is an expert option, that we do not want to expose in the documentation")
+    public static final ConfigOption<Integer> OFFLOAD_SHUFFLE_DESCRIPTORS_THRESHOLD =
+            key("jobmanager.deployment.offload-shuffle-descriptors-threshold")
+                    .intType()
+                    .defaultValue(4_000 * 4_000)
+                    .withDescription(
+                            "The threshold to force enable offload shuffle descriptors via blob server."
+                                    + " This is a fixed value since it is difficult for users to configure."
+                                    + " This default value means JobManager need to serialized and transport"
```

Comment:
`This is a fixed value since it is difficult for users to configure` -> This comments is outdated.